### ISSUE TITLE
custom operator implementation for plaidml

### ISSUE THREAD

**Issue by artyom-beilis** [plaidml/plaidml#1857]
Hello,

I'm looking into adding custom kernels/functions to plaidml.

Today plaidml has nice feature set but its performance is poor in comparison to cuDNN or MIOpen implementations.

I recently started working on dlprimitives project: https://github.com/artyom-beilis/dlprimitives that provides basic functions and inference/training library.

Part of it are core components similar to cuDNN but implemented on OpenCL.

I implemented GEMM and Winograd based convolutions, GEMM, Batch Normalization and many other operators (some of them https://github.com/artyom-beilis/dlprimitives/blob/master/include/dlprim/core_ops.hpp) 

My library already vastly outperforms Keras/Plaidml and Caffe/OpenCL and in many cases gives similar performnace to cudnn/hip based caffe or TF. But dlprimitives library isn't as feature reach. I wonder if there any way to integrate these core custom tuned operators into plaidml so its performance can be boosted to reasonable levels?

Were need I look into?

Thanks!

**Comment by artyom-beilis** [comment]
Anybody? I started working on adding dlprimitives based ops to Caffe-OpenCL but it has very poor memory usage optimization.

I still wonder were to start to getting custom ops to PlaidML?

**Comment by IIIBlueberry** [comment]
Plaidml Op is written in their own EDSL called Tile i found these link, might be helpful to you
[PlaidML Op Tutorial](https://github.com/plaidml/plaidml/wiki/PlaidML-Op-Tutorial/7e0e38161eb89e3317d25bd21fad2930e71376fb)
[Tile Tutorial](https://github.com/plaidml/plaidml/wiki/Tile-Tutorial/61812d1529f54861b6f98656d6357b580155b52d)
[Tile eDSL](https://vertexai-plaidml.readthedocs-hosted.com/en/latest/usage/edsl.html#scope-and-warning)

**Comment by artyom-beilis** [comment]
I understand this. However I want to apply custom - operators that had been highly optimized and not running via generic tile system, since... Manual optimization can win in many cases.
